v2/api/network/customizations/load_balancer_extension.go

```diff
@@ -0,0 +1,171 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT license.
+package customizations
+
+import (
+	"context"
+	"encoding/json"
+	"reflect"
+
+	"github.com/go-logr/logr"
+	"github.com/pkg/errors"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/conversion"
+
+	network "github.com/Azure/azure-service-operator/v2/api/network/v1api20201101storage"
+	. "github.com/Azure/azure-service-operator/v2/internal/logging"
+	"github.com/Azure/azure-service-operator/v2/internal/resolver"
+	"github.com/Azure/azure-service-operator/v2/internal/util/kubeclient"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime"
+	"github.com/Azure/azure-service-operator/v2/pkg/genruntime/extensions"
+)
+
+// Attention: A lot of code in this file is very similar to the logic in route_table_extensions.go and virtual_network_extensions.go.
+// The two should be kept in sync as much as possible.
+
+var _ extensions.ARMResourceModifier = &LoadBalancerExtension{}
+
+func (extension *LoadBalancerExtension) ModifyARMResource(
+	ctx context.Context,
+	armObj genruntime.ARMResource,
+	obj genruntime.ARMMetaObject,
+	kubeClient kubeclient.Client,
+	resolver *resolver.Resolver,
+	log logr.Logger,
+) (genruntime.ARMResource, error) {
+	typedObj, ok := obj.(*network.LoadBalancer)
+	if !ok {
+		return nil, errors.Errorf("cannot run on unknown resource type %T, expected *network.LoadBalancer", obj)
+	}
+
+	// Type assert that we are the hub type. This will fail to compile if
+	// the hub type has been changed but this extension has not been updated to match
+	var _ conversion.Hub = typedObj
+
+	inboundNatRuleGVK := getInboundNatRuleGVK(obj)
```

Comment:
The problem is that we are injecting one type into another. Imagine for example:

LoadBalancer OriginalGVK `20230101`
LoadBalancerNATRule OriginalGVK `20230601`

The "contract" (if you can call it that) is that `20230101` LoadBalancer can embed `20230101` NAT rules, but it can't embed the `20230601` version because that's newer. We need to run conversion.

This has a requirement is poorly documented here and should probably be called out (here and in the vmss and RouteTable extenions too): In each API version, there must be a version of the subresource _and_ the parent resource (no skipping one or the other). We should probably put a comment in azure-arm.yaml to that effect as well next to these resources.

I believe if the GVK doesn't exist, the `List` call will fail saying no such GVK, so at runtime we're "protected" here somewhat.

**The issue is** you can't know statically what GVK you want, so you can't do what you suggest and get the GVK statically. We need to get the GVK dynamically based on GVK that we're going to use to generate the ARM payload.

Reply:
Reiterating this bit:

> This has a requirement that is poorly documented here and should probably be called out (here and in the vmss and RouteTable extenions too): In each API version, there must be a version of the subresource and the parent resource (no skipping one or the other). We should probably put a comment in azure-arm.yaml to that effect as well next to these resources.

My takeaway from @theunrepentantgeek's comment, which we could try to find a way to do, is to ensure that there really _is_ such a version for the parent type and the child type. right now we just make the assumption. Obviously that happens to work in practice but we're not very hardened against issues with that assumption here. Some combination of:
1. Comment in `azure-arm.yaml` saying that there must always be a parent and child in the same API version (to call it out for others when adding new versions)
2. Dynamic checks via scheme lookup here to ensure that such a type actually exists.

Seems the best way out to me?

I at least would be happy to see this done in a separate PR where we tackled all 3-4 of the resources that have this requirement together, rather than trying to do them all here, but also curious what @theunrepentantgeek thinks.

Reply:
Ergh. I get it now. 

My key problem here is that any failure happens at runtime of the controller, while the fix requires changes to the configuration of the generator - these are _very_ far apart in the code base, and the errors will happen at very different times. It's also more than likely the person experiencing the failure won't be the person who needs to change the configuration.

I'd suggest the following attack:
* In this PR, add comments in `azure-arm.yaml` to all the afflicted resources (not just the new resources in this PR).
* Log an issue to track the problem
* Add config + check to the pipeline to enforce the relationship, somehow, so that we can't possibly leave things out
* Revisit this code to see if we can find a less-brittle way to do the conversion that's required